platform/inprocess/mod.rs

```diff
@@ -0,0 +1,359 @@
+                -> Result<(),MpscError>
+    {
+        match self.sender.borrow().send(MpscChannelMessage(data.to_vec(), ports, shared_memory_regions)) {
+            Err(_) => { Err(MpscError::ChannelClosedError) },
+            Ok(_) => { Ok(()) },
+        }
+    }
+}
+
+pub struct MpscReceiverSet {
+    last_index: usize,
+    receiver_ids: Vec<usize>,
+    receivers: Vec<MpscReceiver>,
+}
+
+impl MpscReceiverSet {
+    pub fn new() -> Result<MpscReceiverSet,MpscError> {
+        Ok(MpscReceiverSet {
+            last_index: 0,
+            receiver_ids: vec![],
+            receivers: vec![],
+        })
+    }
+
+    pub fn add(&mut self, receiver: MpscReceiver) -> Result<i64,MpscError> {
+        self.last_index = self.last_index + 1;
+        self.receiver_ids.push(self.last_index);
+        self.receivers.push(receiver.consume());
+        Ok(self.last_index as i64)
+    }
+
+    pub fn select(&mut self) -> Result<Vec<MpscSelectionResult>,MpscError> {
+        let mut receivers: Vec<Option<mpsc::Receiver<MpscChannelMessage>>> = Vec::with_capacity(self.receivers.len());
+        let mut r_id: i64 = -1;
+        let mut r_index: usize = 0;
+
+        {
+            let select = mpsc::Select::new();
+            // we *must* allocate exact capacity for this, because the Handles *can't move*
+            let mut handles: Vec<mpsc::Handle<MpscChannelMessage>> = Vec::with_capacity(self.receivers.len());
+        
+            for r in &self.receivers {
+                let inner_r = mem::replace(&mut *r.receiver.borrow_mut(), None);
+                receivers.push(inner_r);
+            }
+            
+            for r in &receivers {
+                unsafe {
+                    handles.push(select.handle(r.as_ref().unwrap()));
+                    handles.last_mut().unwrap().add();
+                }
+            }
+
+            let id = select.wait();
+            //println!("select id: {} length: {}", id, receivers.len());
+
+            for (index,h) in handles.iter().enumerate() {
+                //println!("index {} id {}", index, id);
+                if h.id() == id {
+                    r_index = index;
+                    r_id = self.receiver_ids[index] as i64;
+                    //println!("r_id {}", r_index);
+                    break;
+                }
+            }
+        }
+
+        // put the receivers back
+        for (index,r) in self.receivers.iter().enumerate() {
+            mem::replace(&mut *r.receiver.borrow_mut(), mem::replace(&mut receivers[index], None));
+        }
+
+        if r_id == -1 {
+            return Err(MpscError::UnknownError);
+        }
+
+        let receivers = &mut self.receivers;
+        match receivers[r_index].recv() {
+            Ok((data, channels, shmems)) => return Ok(vec![MpscSelectionResult::DataReceived(r_id, data, channels, shmems)]),
+            Err(err) => {
+                match err {
+                    MpscError::ChannelClosedError => {
```

Comment:
Could just be folded into the match above: `Err(MpscError::ChannelClosedError) => { ... } Err(_) => { ... }`